# 1. előadás

_TODO: Bevezetés, technikai tudnivalók_

## Helló, világ!

Első programunk a "Helló, világ!" üzenetet jelenti meg a képenyőn. [Régi szokás szerint a legtöbb tankönyv, tananyag ezzel az egyszerű programmal indít.](https://en.wikipedia.org/wiki/%22Hello,_World!%22_program)



In [1]:
print("Helló, világ!")

Helló, világ!


A programunk a ```print``` kulcsszóval kezdődik, ami egy függvénynek a neve. A függvények kerek zárójelpár között várnak meghatározott paramétereket. A ```print``` függvény számára azt kell megadni, hogy mit írjon a kimenetre.

[Bővebben a ```print``` függvényről.](https://www.w3schools.com/python/ref_func_print.asp)

Bármilyen tetszőleges szöveget megadhatunk idézőjelek között:

In [2]:
print("Rendőr, postás, pék is lennék, / kertésznek is vígan mennék, / de leginkább azért főleg / machine learning szakértőnek.")

Rendőr, postás, pék is lennék, / kertésznek is vígan mennék, / de leginkább azért főleg / machine learning szakértőnek.


Természetesen egy program nem csak egy sorból állhat. Lehetőségünk van arra is, hogy egymás után több utasítást adjunk meg. Például több ```print``` függvényt írunk egymás utáni sorokba.

Ilyenkor az egyes utasítások egymás után, fentről lefelé haladva kerülnek végrehajtása. Ezt nevezzük szekvenciális végrehajtásnak. A kimeneten is ennek megfelelően jelennek meg egymás után (egymás alatt) az üzenetek.

Így [Varró Dániel](https://hu.wikipedia.org/wiki/Varr%C3%B3_D%C3%A1niel) versét törtvonalak nélkül, több sorban is megjeleníthetjük:

In [3]:
print("Rendőr, postás, pék is lennék,")
print("kertésznek is vígan mennék,")
print("de leginkább azért főleg")
print("machine learning szakértőnek.")

Rendőr, postás, pék is lennék,
kertésznek is vígan mennék,
de leginkább azért főleg
machine learning szakértőnek.


## Aritmetikai műveletek


A ```print``` függvény segítségével nem csak konstans szövegeket, de matematikai kifejezéseket is lehetőségünk van megjeleníteni. Ilyenkor a fordító kiértékeli a kifejezést, kiszámítja annak eredményét, majd az eredményt fogja a ```print``` függvény a kimeneten megjeleníteni.

A következő 7 alapvető aritmetikai műveletet használhatjuk Python nyelven:

* Összeadás: ```+```

In [4]:
print(2+3)

5


 - Kivonás: ```-```

In [5]:
print(2-3)

-1


 - Szorzás: ```*```

In [6]:
print(2*3)

6


 - Osztás: ```/```

In [7]:
print(2/3)

0.6666666666666666


 - Maradékos osztás: ```%```

Az osztás maradékát fogja megadni. A példában: kettőben a három megvan nulla egésszer, és maradék a kettő.

In [8]:
print(2%3)

2


 - Hatványozás: ```**```

Az ```**``` operátor elé kerül az alap, utána pedig a kitevő. A példában: kettő a harmadik hatványon.

In [9]:
print(2**3)

8


- Egészosztás: ```//```

Megadja lefelé kerekítve az osztás eredményét. A példában: a három a kettőben nulla egésszer van meg.

In [10]:
print(2//3)

0


Lehetőségünk van összetettebb kifejezések megfogalmazására is. A végrehajtás tekintetében a matematikában megszokott szabályok érvényesülnek.

In [11]:
print(2*3-1)

5


In [12]:
print(3**3*3)

81


Zárójelek segítségével tudjuk szabályozni a végrehajtás sorrendjét.

In [13]:
print(2*(3-1))

4


In [14]:
print(3**(3*3))

19683


## Összetett üzenetek kiírása

Lehetőségünk van arra is, hogy a ```print``` segítségével összetett, több üzenetben álló kiírásokat hajtsunk végre. Például szeretnénk megjeleníteni egy szöveget, majd utána egy számítási eredményt is.

In [15]:
print("A kettő és a három szorzatának eredménye:", 2*3)

A kettő és a három szorzatának eredménye: 6


Ilyenkor az egyes üzeneteket vesszővel elválasztva soroljuk fel a zárójelen belül. Az előző példában először egy szöveget, majd egy matematikai kifejezés eredményét jelenítjük meg.

Természetesen kettőnél több paraméter is állhat itt, mindig vesszővel elválasztva egymástól.

In [16]:
print("Ada Lovelace 1815-ben született,", 2023-1815, "évvel ezelőtt. 1852-ben (", 1852-1815, ". évében) hunyt el.")

Ada Lovelace 1815-ben született, 208 évvel ezelőtt. 1852-ben ( 37 . évében) hunyt el.


Látható, hogy a vesszővel elválasztott elemek közé automatikusan egy-egy szóköz kerül a kimeneten. Ezért van az, hogy a ```2023-1815``` kifejezés előtti szövegrész végére és utáni szövegrész elejére nem raktunk szóköz karaktert. Viszont az ```1852-1815``` kifejezés esetén más a helyzet. Itt is láthatjuk az automatikus szóközeket, viszont itt egyik oldalra sem kellene.

Hogyan lehet ezt orvosolni?

A ```print``` függvény rendelkezik több opcionális paraméterrel. Az egyik ilyen a ```sep``` nevezetű, amely a "separator" szó rövidítéséből ered. Ennek segítségével megadható, hogy szóköz helyett milyen karaktert (vagy karaktersorozatot) illesszen be az egyes kiírandó elemek közé. Például megadhatjuk azt, a fenti szöveg esetén, hogy ne szerepeljen semmilyen elválasztó szöveg. Ehhez üres karaktersorozatot (```""```) kell értékül adni a ```sep``` paraméternek a következő módon:

In [17]:
print("Ada Lovelace 1815-ben született,", 2023-1815, "évvel ezelőtt. 1852-ben (", 1852-1815, ". évében) hunyt el.", sep="")

Ada Lovelace 1815-ben született,208évvel ezelőtt. 1852-ben (37. évében) hunyt el.


Látható, hogy az elválasztó szóközök eltűntek. Ezzel a zárójelbe írt életkor problémája megoldódott. Viszont a "208" előtt és után sincs így már szóköz, pedig ott szükség van rá. Ezért egyszerűbb, ha az előtte és utána található szövegrész végére és elejére fűzzük hozzá a kívánt elválasztó szóközöket. Így már megfelelő formában kapjuk meg a szöveget:

In [18]:
print("Ada Lovelace 1815-ben született, ", 2023-1815, " évvel ezelőtt. 1852-ben (", 1852-1815, ". évében) hunyt el.", sep="")

Ada Lovelace 1815-ben született, 208 évvel ezelőtt. 1852-ben (37. évében) hunyt el.


A ```sep``` paraméter értéke bármilyen más tetszőleges szöveg lehet, amit az adott feladat, kívánt cél megkíván. Például kifejezések eredményét pontosvesszővel tagoltan szeretnénk megjeleníteni. Ezt a következő módon tehetjük:

In [19]:
print(2**0, 2**1, 2**2, 2**3, 2**4, 2**5, 2**6, 2**7, sep=";")

1;2;4;8;16;32;64;128


A ```print``` függvény másik opcionális paramétere az ```end```. Ennek is egy karaktert vagy karaktersorozatot adhatunk meg. Ami az ```end``` értéke, azt a szöveget fogja minden kiírás végére elhelyezni a ```print``` függvény. Ahogy a ```sep``` alapértelmezett értéke a szóköz, úgy az ```end``` alapértelmezett értéke a sortörés karakter. Ennek köszönhető, hogy ha egymás után több ```print``` függvényt hívunk meg, akkor az egyes kiírások egymás alatt, és nem egymás mellett fognak megjelenni a kimeneten.

Az előző példa alapján, ha az utolsó kifejezés után is szeretnénk pontosvesszőt megjeleníteni, akkor azt a következő módon tehetjük meg:

In [20]:
print(2**0, 2**1, 2**2, 2**3, 2**4, 2**5, 2**6, 2**7, sep=";", end=";")

1;2;4;8;16;32;64;128;

## Típusok